File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        Set<List<Integer>> set = new HashSet<>();
        List<List<Integer>> ans = new ArrayList<>();

        for (int i = 0; i < nums.length-1; i++) {
            int j = i + 1;
            int k = nums.length - 1;
            while(j<k){
if (nums[i] + nums[j] + nums[k] < 0) {
                j++;
            } else if (nums[i] + nums[j] + nums[k] > 0) {
                k--;
            } else {

                List<Integer> triplet = new ArrayList<>();
                triplet.add(nums[i]);
                triplet.add(nums[j]);
                triplet.add(nums[k]);
                set.add(triplet);

                j++;
                k--;
            }
            }
            
        }

        for (List<Integer> value : set) {
            ans.add(value);
        }

        return ans;
    }
}

File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<vector<int>> ans;
        set<vector<int>> uniquePair;
        for (int i = 0; i < nums.size() - 2; i++) {
            int j = i + 1, k = nums.size() - 1;

            while (j < k) {
                int sum = nums[j] + nums[k];
                if (sum + nums[i] == 0) {
                    uniquePair.insert({nums[i], nums[j], nums[k]});
                    k--;
                    j++;
                } else if (sum + nums[i] < 0) {
                    j++;
                } else {
                    k--;
                }
            }
        }

        for(auto i: uniquePair){
            ans.push_back(i);
        }
        return ans;
    }
};